sourcegen-generator-java/src/main/java/io/micronaut/sourcegen/JavaPoetSourceGenerator.java

```diff
@@ -265,13 +293,36 @@ private void writeRecord(Writer writer, RecordDef recordDef) throws IOException
                 componentBuilder.build()
             );
         }
+
+        addInnerTypes(recordDef.getInnerTypes(), classBuilder, false);
+
         for (MethodDef method : recordDef.getMethods()) {
             classBuilder.addMethod(
                 asMethodSpec(recordDef, method)
             );
         }
-        JavaFile javaFile = JavaFile.builder(recordDef.getPackageName(), classBuilder.build()).build();
-        javaFile.writeTo(writer);
+        return classBuilder;
+    }
+
+    private void addInnerTypes(List<ObjectDef> innerTypes, TypeSpec.Builder classBuilder, boolean isInterface) {
+        for (ObjectDef innerType : innerTypes) {
+            TypeSpec.Builder innerBuilder;
+            if (innerType instanceof ClassDef innerClassDef) {
+                innerBuilder = getClassBuilder(innerClassDef);
+            } else if (innerType instanceof InterfaceDef innerInterfaceDef) {
+                innerBuilder = getInterfaceBuilder(innerInterfaceDef);
+            } else if (innerType instanceof EnumDef innerEnumDef) {
+                innerBuilder = getEnumBuilder(innerEnumDef);
+            } else if (innerType instanceof RecordDef innerRecordDef) {
+                innerBuilder = getRecordBuilder(innerRecordDef);
+            } else {
+                throw new IllegalStateException("Unknown object definition: " + innerType);
+            }
+            if (isInterface) {
+                innerBuilder.addModifiers(Modifier.PUBLIC, Modifier.STATIC);
```

Comment:
`Modifier.PUBLIC` should not be added. There. could be a case when someone wants a private interface.

Is `Modifier.STATIC` required? By default interfaces are static.